src/tensor/diag.jl

```diff
@@ -32,12 +32,13 @@ Base.eltype(::Type{<:Diag{T}}) where {T} = eltype(T)
 Base.convert(::Type{<:Diag{T}},D::Diag) where {T} = Diag{T}(data(D))
 
 Base.similar(D::Diag{T}) where {T} = Diag{T}(similar(data(D)))
+Base.similar(D::Diag{T}, ::Type{T}) where {T} = Diag{T}(similar(data(D)))
 
 # TODO: write in terms of ::Int, not inds
 Base.similar(D::Diag{T},inds) where {T} = Diag{T}(similar(data(D),minimum(dims(inds))))
 Base.similar(D::Type{<:NonuniformDiag{T}},inds) where {T} = Diag{T}(similar(T,length(inds)==0 ? 1 : minimum(dims(inds))))
 
-Base.similar(D::UniformDiag{T}) where {T<:Number} = Diag{T}(zero(T))
```

Comment:
Just like above, it seems like we would still want this version. Was it causing a problem?